Stream implementation (wrapper) for PaginationStream

Ploppz opened the pull request:
## Motivation and Context
https://github.com/awslabs/aws-sdk-rust/discussions/995

## Description
I tried to implement futures::Stream for a wrapper struct around `PaginationStream`. I am unsure if I did it in the best way. After fighting with the borrow checker for a while I decided to try `Arc<Mutex<_>>` - is this the way to go or does there exist a better way?
Even then, does the code look correct? I used it in my project and my integration tests do pass but I am not 100% that these tests will catch any error in paginated ListObjectsV2.

I would appreciate any feedback so far.

## Testing
In progress while waiting for feedback on code

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates


_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

rcoh commented:
for a number of reasons, I think that _is_ the best you can do. Since you need to box the future, you then can't know for sure that the stream is alive long enough.

Luckily, there is a better way! We can expose `poll_next` on `PaginationStream` so that we can delegate to that. I've updated your branch with these changes.

I think there are a couple of remaining items we should do:

1. Add an extension trait so you can do something like:

```rust
s3.list_buckets().into_paginator().send().as_stream_03x().map(...) // and other stream trait useful functions
```

2. Add a changelog entry so folks know this new functionality is available!

Ploppz commented:
Thanks! Some questions:
- Why do you use `tokio-stream` rather than `futures`? I guess I will change the name of the module futures.rs
- What is `03x` in the function name?
- I assume at this point the code is so simply tests are not needed

Ploppz commented:
Does it look good now?

Ploppz commented:
Btw, if tokio-stream re-exports futures::Stream and we don't even use StreamExt for anything, then that means we simply bring in more dependencies than necessary. Shouldn't we just use `futures`?

abusch commented:
Nice, that would be useful to have! In the meantime we've done something pretty similar, which seems to work for us (plus an extension trait to add a `.into_stream()` method):

```rust
pub struct PaginationStreamAdapter<T>(PaginationStream<T>);

impl<T> futures::stream::Stream for PaginationStreamAdapter<T> {
    type Item = T;

    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Option<Self::Item>> {
        let fut = self.0.next();
        pin_mut!(fut);
        fut.poll(cx)
    }
}
```

Ploppz commented:
@abusch I'm not sure but isn't it wrong to call `next()` in every call to `poll_next`? 
Even if it somehow works, it seems wrong to me. I might be wrong though but here's my line of thought:
The async system will keep polling to check the status of whether the next item is ready, and every time it does that, internally, your impl will call the next() function which I believe (these semantics would make sense to me) should be waited for completion until it's called the next time. If you are lucky, it's always ready when polled, maybe that's why it works for you?

rcoh commented:
FYI I'm pretty sure that implementation is wrong — It may work due to a
coincidence of how our demand driven paginator works but I wouldn't
recommend relying on that

On Thu, Dec 14, 2023, 2:35 AM Antoine Büsch ***@***.***>
wrote:

> Nice, that would be useful to have! In the meantime we've done something
> pretty similar, which seems to work for us (plus an extension trait to add
> a .into_stream() method):
>
> pub struct PaginationStreamAdapter<T>(PaginationStream<T>);
> impl<T> futures::stream::Stream for PaginationStreamAdapter<T> {
>     type Item = T;
>
>     fn poll_next(
>         mut self: Pin<&mut Self>,
>         cx: &mut std::task::Context<'_>,
>     ) -> std::task::Poll<Option<Self::Item>> {
>         let fut = self.0.next();
>         pin_mut!(fut);
>         fut.poll(cx)
>     }}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/smithy-lang/smithy-rs/pull/3299#issuecomment-1855324628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADYKZ37BBIWTBU75TEYIU3YJKT3ZAVCNFSM6AAAAABAMQI67CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJVGMZDINRSHA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

abusch commented:
Hmm, yeah, you might be right 😓

rcoh commented:
yeah that's fair. In that case, I think we should use `future_core` stream directly: [https://docs.rs/futures-core/latest/futures_core/stream/trait.Stream.html](futures_core::Stream) (futures::Stream) is just yet another re-export of this.

We should probably include a comment suggesting that tokio_stream is where most of the useful utilities live, though

rcoh commented on rust-runtime/aws-smithy-types-convert/src/stream.rs:
```diff
@@ -0,0 +1,52 @@
+}
+
+impl<Item> PaginationStreamExt<Item> for PaginationStream<Item> {
+    fn as_stream_01x(self) -> PaginationStreamImplStream<Item> {
```
I wonder if this should be `into` instead since it consumes `self`. I don't have a strong opinion.

rcoh commented on CHANGELOG.next.toml:
```diff
@@ -79,3 +79,9 @@ message = "Expose local socket address from ConnectionMetadata."
 meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client" }
 author = "declanvk"
+
+[[smithy-rs]]
+message = "Optional implementation of Stream on PaginationStream"
+references = ["smithy-rs#3299"]
+meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client"}
+author = "Erlend Langseth <erlend.langseth@openanalytics.eu>"
```
we should probably also duplicate this under `aws-sdk-rust`, this will be relevant for folks using the SDK as well.

I think we should also change the message to be a little clearar about what folks can get and how, something like:

---

Add `PaginationStreamExt` extension trait to `aws-smithy-types-convert` behind the `convert-streams` feature. This makes it possible to treat a paginator as a [`futures_core::Stream`](https://docs.rs/futures-core/latest/futures_core/stream/trait.Stream.html), allowing customers to use stream combinators like [`map`](https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.map) and [`filter`](https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.filter).

Example: (double check this compiles)
```rust
use aws_smithy_types_convert::stream::PaginationStreamExt
s3_client.list_objects_v2().bucket("...").send().await.into_paginator().items().into_stream_03x()
```

---

Ploppz replied:
can I write all that in `message` with markdown formatting? Or maybe I can add that code example in rustdocs instead?

rcoh replied:
yep, use triple quotes to surround your message `"""` and paste it in.

rcoh replied:
not sure if you can easily grab the markdown out of my message, if not here: https://gist.github.com/rcoh/b69e8a42509842e239cd460b68053e16

Ploppz replied:
thanks, but a trick is to quote your message to get the markdown ;)

rcoh reviewed (commented):
this is looking great! just a couple of suggestions then we'll get this merged. Thanks for working on this, I think customers are really going to appreciate it.

Ploppz commented:
I implemented the suggested changes, and rebased onto latest upstream `main` branch to fix conflicts. Also removed a trait bound that was unnecessary. Couldn't write very helpful code in the rustdoc unfortunately because we don't have a lot of dependencies to work with there.

rcoh commented on CHANGELOG.next.toml:
````diff
@@ -109,3 +109,31 @@ message = "Add support for constructing [`SdkBody`] and [`ByteStream`] from `htt
+```
+"""
+references = ["smithy-rs#3299"]
+meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client"}
````
```suggestion
meta = { "breaking" = false, "tada" = false, "bug" = false }
```

rcoh commented on CHANGELOG.next.toml:
```diff
@@ -109,3 +109,31 @@ message = "Add support for constructing [`SdkBody`] and [`ByteStream`] from `htt
+"""
+references = ["smithy-rs#3299"]
+meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client"}
+author = "Erlend Langseth <erlend.langseth@openanalytics.eu>"
```
```suggestion
author = "Ploppz"
```

rcoh commented on CHANGELOG.next.toml:
```diff
@@ -109,3 +109,31 @@ message = "Add support for constructing [`SdkBody`] and [`ByteStream`] from `htt
+"""
+references = ["smithy-rs#3299"]
+meta = { "breaking" = false, "tada" = false, "bug" = false }
+author = "Erlend Langseth <erlend.langseth@openanalytics.eu>"
```
```suggestion
author = "Ploppz"
```
